Firestore/Swift/Source/PropertyWrapper/FirestoreQueryObservable.swift

```diff
@@ -0,0 +1,162 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import SwiftUI
+import FirebaseFirestore
+
+@available(iOS 14.0, *)
+@available(tvOS, unavailable)
+internal class FirestoreQueryObservable<T>: ObservableObject {
+  @Published var items: T
+
+  private let firestore = Firestore.firestore()
+  private var listener: ListenerRegistration?
+
+  private var setupListener: (() -> Void)!
+
+  internal var configuration: FirestoreQuery<T>.Configuration {
+    didSet {
+      removeListener()
+      setupListener()
+    }
+  }
+
+  init<U: Decodable>(configuration: FirestoreQuery<T>.Configuration) where T == [U] {
+    items = []
+    self.configuration = configuration
+    setupListener = createListener { [weak self] querySnapshot, error in
+      if let _ = error {
+        // in this scenario, we actively ignore the error
+        self?.items = []
+        return
+      }
+
+      guard let documents = querySnapshot?.documents else {
+        self?.items = []
+        return
+      }
+
+      self?.items = documents.compactMap { document in
+        try? document.data(as: U.self)
+      }
+    }
+
+    setupListener()
+  }
+
+  init<U: Decodable>(configuration: FirestoreQuery<T>.Configuration) where T == [Result<U, Error>] {
+    items = []
+    self.configuration = configuration
+    setupListener = createListener { [weak self] querySnapshot, error in
+      if let error = error {
+        self?.items = [.failure(error)]
+        return
+      }
+
+      guard let documents = querySnapshot?.documents else {
+        self?.items = []
+        return
+      }
+
+      self?.items = documents.compactMap { queryDocumentSnapshot in
+        Result {
+          // NOTE: The `try` handles the parse failure
+          // The `Optional` signature for the result is a
+          // workaround for when a document may not exist.
+          // - which they always do when looping over
+          // them in the `documents` property of a snapshot
+          try queryDocumentSnapshot.data(as: U.self)!
+        }
+      }
+    }
+
+    setupListener()
+  }
+
+  init<U: Decodable>(configuration: FirestoreQuery<T>.Configuration) where T == Result<[U], Error> {
+    items = .success([])
+    self.configuration = configuration
+    setupListener = createListener { [weak self] querySnapshot, error in
+      if let error = error {
+        self?.items = .failure(error)
+        return
+      }
+
+      guard let documents = querySnapshot?.documents else {
+        self?.items = .success([])
+        return
+      }
+
+      do {
+        let items = try documents.compactMap { queryDocumentSnapshot in
+          try queryDocumentSnapshot.data(as: U.self)
+        }
+        self?.items = .success(items)
+      } catch {
+        self?.items = .failure(error)
+      }
+    }
+
+    setupListener()
```

Comment:
Do we have a canonical way to implement snapshot listeners in Swift, @morganchen12 and @schmidt-sebastian? In [my article](https://peterfriese.dev/firestore-codable-the-comprehensive-guide/), I've tried to condense what we've got in the documentation, but some aspects aren't documented very clearly.

Reply:
We do not, though the method in your article feels a lot like the culmination of all of the scattered recommendations in our docs. Which aspects are you referring to in particular?